File: code/algo/getoffer/go/3/uniquearr_all_subset.go
package main

func permute(nums []int) [][]int {
	n := len(nums)
	var ans [][]int
	var path []int
	vis := make([]bool, n)
	var dfs func()
	dfs = func() {
		if len(path) == n {
			ans = append(ans, append([]int{}, path...))
			return
		}
		for i := 0; i < n; i++ {
			if !vis[i] {
				vis[i] = true
				path = append(path, nums[i])
				dfs()
				vis[i] = false
				path = path[:len(path)-1]
			}
		}
	}
	dfs()
	return ans
}

File: code/algo/getoffer/c++/3/uniquearr_all_subset.cpp
#include <vector>

using namespace std;

void dfs(vector<int> &nums, vector<bool> &vis, vector<int> &path,
         vector<vector<int>> &ans) {
  if (path.size() == nums.size()) {
    ans.push_back(path);
    return;
  }
  for (int i = 0; i < nums.size(); ++i) {
    if (!vis[i]) {
      vis[i] = true;
      path.push_back(nums[i]);
      dfs(nums, vis, path, ans);
      vis[i] = false;
      path.pop_back();
    }
  }
}

vector<vector<int>> permute(vector<int> &nums) {
  vector<vector<int>> ans;
  vector<int> path;
  vector<bool> vis(nums.size(), false);
  dfs(nums, vis, path, ans);
  return ans;
}
